### PR TITLE
Context propagation refactor

### PR DESCRIPTION
**Pull request by Blacksmoke16**
Partially/fully resolves:

* https://github.com/open-telemetry/opentelemetry-php/issues/239
* https://github.com/open-telemetry/opentelemetry-php/issues/266
* https://github.com/open-telemetry/opentelemetry-php/issues/240

Are still some things that need done to be 100% compliant. I'll go thru the changes again and get a more detailed list, but the main two are:

* Moving most of the new logic into the API side of things
  * Given our current API implementation is only interfaces I figured it would be better to save this for later
* Refactor `Tracer` to change how spans can be created and activated in one method
  * Possibly via `Tracer#inSpan` that takes a callable?
  * Is slightly painful since you need to return both the span and the scope

TODO:

- [x] Document what still needs to be updated to be spec compliant
- [x] Add some doc comments where applicable to new types/methods
- [x] Better unit test coverage of new logic
- [ ] TBD

Future PR work:

* Fully implement https://github.com/open-telemetry/opentelemetry-php/issues/266
  * Do a pass on Span API to ensure the methods on the interface match spec
  * Maybe save implementing it in the API side for later?
* Decouple `Baggage` from `Context`
  * Make it its own class that implements the `Baggage` and `ImplicitContextKeyed` interfaces
* Refactor `Tracer`
  * Should no longer handle keeping track of active spans/making them active
    * https://github.com/open-telemetry/opentelemetry-specification/blob/v1.6.1/specification/trace/api.md#tracer
* Move the implementation of some things from the SDK to API side of things
  * Mainly noop span and context storage logic
    * Probably via abstract classes in the case of PHP?
* Support propagation of `Baggage`
  * Create a new `TextMapPropagator`
  * Create a new `CompositeTextMapPropagator` type to wrap Trace and Baggage implementations
    * https://github.com/open-telemetry/opentelemetry-specification/blob/v1.6.1/specification/context/api-propagators.md#composite-propagator
  * Move types to the API layer and expose global propagator types
    * https://github.com/open-telemetry/opentelemetry-specification/blob/v1.6.1/specification/context/api-propagators.md#global-propagators

### CODE REVIEW

File: Context/Context.php [hunk 1]
```diff
@@ -4,9 +4,6 @@
 
 namespace OpenTelemetry\Context;
 
-/**
- * @template TContext of Context
```
Comment (Blacksmoke16) [inline Context/Context.php:8]: Going to look into if I can bring this back, but psalm was complaining when trying to pass a context instance to `TraceContextPropagator#inject`.

File: Context/Context.php [hunk 2]
```diff
@@ -64,6 +69,22 @@ public function set(ContextKey $key, $value)
         return new static($key, $value, $this);
     }
 
+    /**
+     * @todo: Implement this on the API side
+     */
+    public function with(ImplicitContextKeyed $value): Context
```
Comment (Blacksmoke16) [inline Context/Context.php:75]: Given `Context` is an immutable collection I went with `with`, but given PHP doesn't have overloads I'm open to other names. I considered `withSpan` but this can also be used with baggage :/

Additionally would make it make sense to update `setValue` to `withValue` as I think that's a more common name for immutable containers.

File: Context/Context.php [hunk 3]
```diff
@@ -94,19 +115,16 @@ public static function setValue(ContextKey $key, $value, $parent=null)
...
      */
     public function get(ContextKey $key)
     {
         if ($this->key === $key) {
             return $this->value;
         }
+
         if (null === $this->parent) {
-            throw new ContextValueNotFoundException();
+            return null;
```
Comment (Blacksmoke16) [inline Context/Context.php:127]: This relates to https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/error-handling.md. Allows returning an invalid span in the case of there not being an active span. Checking for `null` seemed better than catching an exception.

File: Context/ImplicitContextKeyed.php [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+declare(strict_types=1);
+
+namespace OpenTelemetry\Context;
+
+/**
+ * Represents a value that can be sored within {@see Context}.
+ * Allows storing themselves without exposing a {@see ContextKey}.
+ */
+interface ImplicitContextKeyed
```
Comment (Blacksmoke16) [inline Context/ImplicitContextKeyed.php:11]: This mirrors the Java implementation and allows for `Baggage` to follow the same patterns as `Span`. Java apparently allows defining default method implementations on interfaces, so there is some minor duplication, but could introduce a trait similar to what `ContextValueTrait` was doing if it gets out of hand.

File: Makefile [hunk 5]
```diff
@@ -38,7 +38,7 @@ proto:
 bash:
 	$(DC_RUN_PHP) bash
 style:
-	$(DC_RUN_PHP) php ./vendor/bin/php-cs-fixer fix --config=.php-cs-fixer.php --stop-on-violation --using-cache=no -vvv
+	$(DC_RUN_PHP) php ./vendor/bin/php-cs-fixer fix --config=.php-cs-fixer.php --using-cache=no -vvv
```
Comment (Blacksmoke16) [inline Makefile:41]: Unrelated, but this was annoying me as you needed to run `make style` multiple times just to fix everything.

File: api/Trace/Span.php [hunk 6]
```diff
@@ -55,5 +56,7 @@ public function setSpanStatus(string $code, ?string $description = null): Span;
      */
     public function end(int $timestamp = null): Span;
 
+    public function getSpanContext(): SpanContext;
```
Comment (Blacksmoke16) [inline api/Trace/Span.php:59]: This is defined no the Java Span API, and is needed due to where active spans and such are stored now. E.g. within `TraceContextPropagator::inject`. I did also notice there is `Span#getContext()` that returns a clone of the span context, possibly could just replace that method with this one.

File: api/Trace/TextMapPropagator.php [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+declare(strict_types=1);
+
+namespace OpenTelemetry\Trace;
+
+use OpenTelemetry\Context\Context;
+
+/**
+ * @see https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/context/api-propagators.md#textmap-propagator
+ */
+interface TextMapPropagator
```
Comment (Blacksmoke16) [inline api/Trace/TextMapPropagator.php:12]: This used to be `TextMapFormatPropagator`. Renamed and update interface based on latest spec.

File: api/Trace/TextMapPropagator.php [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+    /**
+     * Injects specific values from the provided {@see Context} into the provided carrier
+     * via an {@see PropagationSetter}.
+     *
+     * @see https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/context/api-propagators.md#textmap-inject
+     *
+     * @param mixed $carrier
+     */
+    public static function inject(&$carrier, Context $context = null, PropagationSetter $setter = null): void;
```
Comment (Blacksmoke16) [inline api/Trace/TextMapPropagator.php:31]: I took some inspiration from the Ruby implementation where it made the context and setter optional to fall back on current context and the default setter (`PropagationMap` in our case).

Just makes it a bit easier to use, but happy to not do that if we'd rather be more explicit.

File: sdk/Baggage.php [hunk 9]
```diff
@@ -4,23 +4,20 @@
...
 namespace OpenTelemetry\Sdk;
 
+use Generator;
 use OpenTelemetry\Context\Context;
 use OpenTelemetry\Context\ContextKey;
 
-/**
- * @extends Context<Baggage>
- */
 class Baggage extends Context
```
Comment (Blacksmoke16) [inline sdk/Baggage.php:11]: Ultimately I think it would make more sense to better follow Java's implementation and make `Baggage` _not_ extend `Context`, but instead implement `ImplicitContextKeyed` itself.

That'll be for another PR tho.

File: sdk/Trace/NoopSpan.php [hunk 10]
```diff
@@ -4,17 +4,21 @@
...
 use Throwable;
 
-class NoopSpan implements API\Span, ReadableSpan
+/**
+ * @see https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#wrapping-a-spancontext-in-a-span
+ * @todo: Implement this on the API side.
+ * @todo: Can we just use {@see https://www.php.net/manual/en/language.oop5.anonymous.php}?
+ * @todo: If not rename this to `NonRecordingSpan`.
+ */
+class NoopSpan implements ReadWriteSpan
```
Comment (Blacksmoke16) [inline sdk/Trace/NoopSpan.php:20]: Still some work to be done here. 

Ref: https://github.com/open-telemetry/opentelemetry-php/issues/266

File: sdk/Trace/NoopSpan.php [hunk 11]
```diff
@@ -53,6 +56,7 @@ public function __construct(API\SpanContext $spanContext = null)
         }
         $this->attributes = new Attributes();
         $this->events = new Events();
+        $this->links = new Links();
```
Comment (Blacksmoke16) [inline sdk/Trace/NoopSpan.php:59]: Missed this in #408 and #406.

File: sdk/Trace/Span.php [hunk 12]
```diff
@@ -47,6 +47,44 @@ class Span implements API\Span, ReadableSpan
...
+    }
+
+    public static function getInvalid(): NoopSpan
+    {
+        if (null === self::$invalidSpan) {
+            self::$invalidSpan = new NoopSpan();
+        }
+
+        return self::$invalidSpan;
+    }
```
Comment (Blacksmoke16) [inline sdk/Trace/Span.php:77]: Kinda annoying needing to have a static var just for this, but is needed so it returns the same instance given we cant just define it as a PHP const. Maybe could via `define`, but at that point would be easier to have a dedicated type and just have this method delegate to that.

File: sdk/Trace/TraceContextPropagator.php [hunk 13]
```diff
@@ -16,40 +22,55 @@
  * traceparent header and relevant parts of the tracestate header containing
  * their proprietary information.
  */
-final class TraceContext implements API\TextMapFormatPropagator
+final class TraceContextPropagator implements API\TextMapPropagator
```
Comment (Blacksmoke16) [inline sdk/Trace/TraceContextPropagator.php:25]: Renamed this to make it more clear what it is, and updated implementation to better follow the new interface. Used https://github.com/open-telemetry/opentelemetry-java/blob/main/api/all/src/main/java/io/opentelemetry/api/trace/propagation/W3CTraceContextPropagator.java as a reference.

File: sdk/Trace/TraceContextPropagator.php [hunk 14]
```diff
@@ -16,40 +22,55 @@
...
 
-    /**
-     * {@inheritdoc}
-     */
-    public static function inject(API\SpanContext $context, &$carrier, API\PropagationSetter $setter): void
+    /** {@inheritdoc} */
+    public static function inject(&$carrier, Context $context = null, PropagationSetter $setter = null): void
     {
+        $context = $context ?? Context::getCurrent();
+        $setter = $setter ?? new PropagationMap();
```
Comment (Blacksmoke16) [inline sdk/Trace/TraceContextPropagator.php:41]: Need to read the spec a bit more, but might be better to have these be singletons and expose them via class methods.

File: sdk/Trace/Tracer.php [hunk 15]
```diff
@@ -124,6 +124,10 @@ public function setActiveSpan(API\Span $span): void
         $this->tail[] = $this->active;
 
         $this->active = $span;
+
+        // FIXME: This should either be called manually or as part of a dedicated tracer operation,
+        // probably some sort of `inSpan` method?
+        $span->makeCurrent();
```
Comment (Blacksmoke16) [inline sdk/Trace/Tracer.php:130]: This relates to the 2nd major thing that needs to be done. Essentially this should either be called manually by the user, or we have an `inSpan` method that makes a yielded span active for the duration of a callback. Otherwise it might be hard to keep a dedicated method that creates a span and makes it the current one since it would need to return _both_ the span itself and a `Scope` to reset the active span.

File: sdk/Trace/Span.php [hunk 16]
```diff
@@ -334,12 +372,19 @@ public function isStatusOk(): bool
...
+    public function storeInContext(Context $context): Context
+    {
+        return $context->set(SpanContextKey::instance(), $this);
+    }
+
+    /**
+     * @todo: Implement this on the API side
      */
-    protected static function getContextKey(): ContextKey
+    public function makeCurrent(): Scope
```
Comment (Blacksmoke16) [inline sdk/Trace/Span.php:385]: This is what Java called it, but `makeActive` would might make more sense. Open to other suggestions.

File: Context/Context.php (hunk 2 above)
Comment (bobstrecansky) [reply Context/Context.php:75]: With seems adequate to me for now.  I agree with you that `withSpan` may get confusing alongside baggage.

`withValue` also makes sense to me for immutable containers.  

Naming these attributes is difficult.  What synonym would be applicable with `Value` in this particular instance 🤔

File: Context/ImplicitContextKeyed.php (hunk 4 above)
Comment (bobstrecansky) [reply Context/ImplicitContextKeyed.php:11]: Sounds like a plan to me.  Following the JAVA spec is perfectly acceptable here!

File: Makefile (hunk 5 above)
Comment (bobstrecansky) [reply Makefile:41]: No problem.  Happy this helps with the developer experience!

File: api/Trace/Span.php (hunk 6 above)
Comment (bobstrecansky) [reply api/Trace/Span.php:59]: Yeah replacing `Span#getContext()` with this sounds a bit cleaner to me.

File: api/Trace/TextMapPropagator.php (hunk 8 above)
Comment (bobstrecansky) [reply api/Trace/TextMapPropagator.php:31]: Agreed that this makes things a bit easier to use; I'm all for it.

File: sdk/Baggage.php (hunk 9 above)
Comment (bobstrecansky) [reply sdk/Baggage.php:11]: Sounds good to me.  Maybe worth making an issue about it?

File: sdk/Trace/TraceContextPropagator.php (hunk 14 above)
Comment (bobstrecansky) [reply sdk/Trace/TraceContextPropagator.php:41]: I think you're correct here - I can't think of cases where we'd have more than one setter.

File: sdk/Trace/Span.php (hunk 16 above)
Comment (bobstrecansky) [reply sdk/Trace/Span.php:385]: I'd have to think back, but I don't think there is a way to make a span current but not active.  `makeActive` feels like the sane choice to me here too.

Comment (Blacksmoke16) [reply sdk/Trace/Span.php:385]: I also thought of `->activate()` if we think that's better 🤷. That seems to be what C# uses.

File: api/Trace/Span.php (hunk 6 above)
Comment (Blacksmoke16) [reply api/Trace/Span.php:59]: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#get-context

Looks like we should do the opposite, my as well use the name the spec suggests.

File: sdk/Baggage.php (hunk 9 above)
Comment (Blacksmoke16) [reply sdk/Baggage.php:11]: Was planning on tackling Baggage after getting context/propagation up to date. But I can create a more generic Baggage issue if you want it for tracking?

File: sdk/Trace/TextMapGetterSetter.php [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+use function sprintf;
+use function strtolower;
+
+/**
+ * @see https://github.com/open-telemetry/opentelemetry-specification/blob/v1.6.1/specification/context/api-propagators.md#textmap-propagator Getter and Setter.
+ *
+ * Default implementation of {@see API\PropagationGetter} and {@see API\PropagationSetter}.
+ * This type is used if no custom getter/setter is provided to {@see API\TextMapPropagator::inject()} or {@see API\TextMapPropagator::extract()}.
+ */
+class TextMapGetterSetter implements API\PropagationGetter, API\PropagationSetter
```
Comment (Blacksmoke16) [inline sdk/Trace/TextMapGetterSetter.php:25]: This used to be called `PropagationMap` but I renamed it to be a bit more clear on what it actually is.

File: sdk/Trace/TextMapGetterSetter.php [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+            self::$instance = new self();
+        }
+
+        return self::$instance;
+    }
+
+    /** {@inheritdoc} */
+    public function keys($carrier): array
+    {
+        if (is_array($carrier)) {
```
Comment (Blacksmoke16) [inline sdk/Trace/TextMapGetterSetter.php:45]: This method is required by the API but was missing an implementation. Opted to only support arrays even tho the actual getter supports `ArrayAccess` as that interface doesn't expose a way to get all the keys so oh well.

File: sdk/Trace/TextMapGetterSetter.php (hunk 17 above)
Comment (Blacksmoke16) [reply sdk/Trace/TextMapGetterSetter.php:25]: Although maybe `ArrayGetterSetter` would be even better...

File: sdk/Trace/Tracer.php (hunk 15 above)
Comment (bobstrecansky) [reply sdk/Trace/Tracer.php:130]: The Java pattern seems reasonable to me.  Leaving those details to the user might make things more flexible in the long run.  We can always update it later if we get complaints / don't love the implementation details.

File: sdk/Trace/Span.php (hunk 16 above)
Comment (bobstrecansky) [reply sdk/Trace/Span.php:376]: `->activate()` makes sense to me - that feels explicit enough.

File: sdk/Trace/TextMapGetterSetter.php (hunk 17 above)
Comment (bobstrecansky) [reply sdk/Trace/TextMapGetterSetter.php:25]: There aren't going to be times when we are GetterSetter'ing anything besides an array, correct?

File: sdk/Trace/TextMapGetterSetter.php (hunk 18 above)
Comment (bobstrecansky) [reply sdk/Trace/TextMapGetterSetter.php:45]: Yup; I like that plan.

File: sdk/Trace/TextMapGetterSetter.php (hunk 17 above)
Comment (Blacksmoke16) [reply sdk/Trace/TextMapGetterSetter.php:25]: I don't think we can have something for _all_ use cases. The majority of the time the carrier is most likely going to be an array, so having that be the default is fine. If that is not the case for someone, then they would need to define their own getter/setter type. I.e. there is a dedicated object or something.

Might make sense to make the carrier the 2nd argument as that might be more common than using a custom context?

File: sdk/Trace/TextMapGetterSetter.php (hunk 18 above)
Comment (Blacksmoke16) [reply sdk/Trace/TextMapGetterSetter.php:45]: Option 2: Define another interface that extends `ArrayAccess` to expose the `keys`, then only accept that or array. This way we can still call it `ArrayGetterSetter` but support both built in arrays and array-like objects. Thoughts?

File: Context/Context.php (hunk 2 above)
Comment (Blacksmoke16) [reply Context/Context.php:78]: What about `withContextValue`? I.e. _a value that is meant to be stored in context._ This means the API would be:

* `set` => `with`
* `setValue` => `withValue`
* `with` => `withContextValue`

Comment (bobstrecansky) [reply Context/Context.php:78]: Sounds good to me!  Thanks for the clear, concise explanation.

File: sdk/Trace/TextMapGetterSetter.php (hunk 17 above)
Comment (bobstrecansky) [reply sdk/Trace/TextMapGetterSetter.php:25]: Agreed that a carrier will certainly be more common than using a custom context.  Sounds like a plan to me.

File: sdk/Trace/TextMapGetterSetter.php (hunk 18 above)
Comment (Blacksmoke16) [reply sdk/Trace/TextMapGetterSetter.php:45]: What about `ArrayAccessGetterSetter` then it's clear it could be used for anything that can be accessed as an array?

Comment (bobstrecansky) [reply sdk/Trace/TextMapGetterSetter.php:45]: Sounds like a plan!

**Review by bobstrecansky — APPROVED** [review]
